### DISCUSSION TITLE
Some Barotrauma Suggestions -- QoL Features, Fixes, Desiderata

### DISCUSSION THREAD

**Discussion by Arctodites** [FakeFishGames/Barotrauma#9692] [Ideas and suggestions]
This is a post for ideas and suggestions that I hope the dev team might consider for future releases.  I'll update it as new things come to mind, and I'd welcome any ideas that fellow players want to add.

[Note: Updated 13, 19 Aug. for formatting and additional content]

-----
**1.) WALLS and STRUCTURES:**
1.1) Shells and Hulls:
- Storage containers, crowbars, etc. shouldn't be able to bash through interior walls so easily, especially when they can't bash through doors
- 'Shell A' pieces could be further classified as either 'straight' or 'curved,' and grouped in the entity menu accordingly.
- It would be nice to have straight 'Shell A' pieces for each angle, in addition to the curved versions
- Retractable half-round or quarter-round domed shell sections -- a glass version and a metal version
- A wider variety of art styles for the shells of the main subs in addition to 'Shell A': a dark, subtle pattern (more akin to the shuttle shell pieces) whose texture tiles seamlessly, and maybe some exotic shells: e.g., ice, scale, bone, etc.

1.2) Interior Walls:
- All 'Large Inside Wall Cap' pieces should comply with their sprite depth in relation to other shells and walls.  Currently it is not possible to obstruct a wall cap with a 'Shell A' piece, for example, unless one sets the sprite depth of the wall cap to > 0.5
- There should be 15 deg., 30 deg., 45 deg., and 60 deg. 'Large Inside Wall' pieces with the same artwork as the existing 'Large Inside Wall' sprites

1.3) Doors and Hatches:
- A wider selection of doors and gates, including half-height gates that obstruct movement, but which let through light, air, and water
-  When prying shut a door or a hatch, the text on the progress bar should read 'Closing...,' not 'Opening...'  Otherwise, a neutral term such as 'Forcing...' should be used.
- Option to disable the message: "This door is controlled by buttons or remote controllers wired to it."  This reminder is especially annoying for custom subs in which one chooses not to use the prefab 'Door with Buttons' or to make doors non-interactable
- Background shutters or vertical roller doors (of several form factors) that can be opened and closed (like the 'curtain' item) to hide / reveal whatever is behind them (e.g., display cases, equipment installations, etc.) and (preferably) to make the objects behind them non-interactable when the shutters are closed;  a lockable version of such doors or shutters would also be useful
- 30 deg., 45 deg., and 60 deg. doors, hatches, and docking hatches

-----
**2.) LABELS, ICONS, and SIGNS:**
2.1) Label Icons:
- It would be useful to have label icons for the following:
-- Oxygen (e.g., a stylised 'O^2' icon); 
-- Lighting (e.g., a lightbulb icon)
-- Warning (e.g., exclamation point inside a triangle)
-- Ship Weapons Systems (e.g., same icon as that used in the Status Monitor HUD; also the targeting reticle icon used in the Security Officer tutorial campaign)
-- Monster (e.g., same as tentacles icon used for 'Report Intruders')
-- Tool(s) (e.g., crossed pair of wrench and screwdriver OR welding tool icon that is already used for the Mechanic tutorial)
-- Machine / Maintenance (e.g., a toothed gear)
-- Materials (e.g., a stylised 'Carbon' sprite)
-- Door / Exit (e.g., same icon already used for yellow/black door sign and in the campaign tutorials)

-----
**3.) ITEMS AND STRUCTURES (generally):**
3.1) Displays, Indicators, Instruments
- A resizable (i.e., infinitely extensible) version of the 'Hologram Display.' (I like to set the 'Text Display' item's alpha to zero and use the hologram display as a background for the text instead)
- Digital bargraph meters: circular and edgewise configurations with three or five bars; monocolour and tricolour

3.2) Text Displays:
- Ability to centre or right-align the text on a display
- Ability to vertically centre the text on a display (especially when the text is resized)
- A text display item with a wider (i.e., longer) form factor
- A text display with a shorter form factor, capable of displaying a maximum of 3 or 4 characters (e.g., for use with counters)
- Ability to chain text displays together, such that the the texts across all displays scroll in synchronisation with each other
- (Thanks to Ek-Videogames for this one): a choice of fonts with better resolution / scalability and kerning for legibility

3.3) Hand Weapons
- Sprites for many guns incl. 'Riot Shotgun' and 'SMG' are very low-res and blurry up-close (e.g., in the 'Weapon Holder' slot); same for 'Fire Extinguisher' sprites.
- Shooting at fire extinguishers, welding tanks, or oxygen tanks should cause them to explode.

3.4) Housings, Casings, and Trimwork
- A variety of styles and form factors of (purely cosmetic) housings for equipment, lights, components, cabinets, etc.  There are plenty of everyday decorative items out of which to make such housings already, but it often involves using a lot of separate pieces to make it look cool.

3.5) Onboard Machines, Equipment, etc.
- A cryostasis chamber in which severely injured / infected / unruly crewmembers can be frozen (but not without risk of adverse consequences)
- A generic (human) version of the Alien terminal, into which any specified small item may be placed to act as a key

3.6) Cabinets and Other Storage Furniture
- Diving suit lockers, ammo loader bases (incl. depth charge loaders), oxygen generators, and oxygen tank shelves should have a binary data output pin that signals whether they contain (at least one) item or not: 0 = empty, 1 = contains at least one item
- A small version of the secure steel cabinet with only three item slots
- A generic, three-slot 'Compartment' storage item that is not named 'Medical Compartment'
- A three-slot 'Tool Holder' similar to the 'Weapon Holder,' which displays the tools held in it
- Functioning display cases, trophy cases, mounts, stands, etc. on which to display items and collectibles

3.7) Stairs, Ramps, Lifts, etc.
- Ramps with a range of slopes, including  a slope that is less steep than stairs; ideally such ramps could be not only scalable, but extensible
- More sizes and form factors for stairs
- Cargo lift / elevator: even if you create a system of pneumatic delivery tubes for the subs, elevator platforms or cargo lifts would still be a desirable feature
- Moving conveyor belts
- Pneumatic or hydraulic pistons (vertical and horizontal versions) (e.g., for pushing cargo, lifting platforms, or squishing monsters)

3.8) Background Walls and Structures
- More styles of railings, fences, and parapets, preferably extensible ones

3.9) Miscellaneous
- Rechargeable, convex forcefields / plasma shields (both personal shields and for ships / shuttles)
- Vending machines that require power and that dispense only items pre-selected in the submarine editor
- A vertical version of the yellow-and-black cordon (i.e., 'Construction Barrier')

3.10) Entity properties
- In addition to the 'Msg' field in the entity properties of the sub editor, there should be a general 'name' or 'description' field that can be used for naming specific entities: e.g., the a 'Small Pump' could be named 'fire suppression pump,' and this name would show up directly beneath 'Small Pump' in the in-game pop-up, preferably in a unique colour (viz., not the same shade of yellow or blue as the other text)

-----
**4.) WIRING COMPONENTS:**
4.1) Components
- A dedicated NAND component
- A three-pin version of the AND, OR, XOR, and ADDER components

4.2) Buttons, Levers, Switches, Alarms
- A greater variety of button styles (e.g., a simple round button; a simple square button; a button with a wide (i.e., horizontal) form factor)
- Switches with only two states (on / off, i.e. 1/0) both of whose indicator lights (red / green) work regardless of whether "IsToggle" is checked
- Switches with different indicator light colour schemes (e.g., blue/yellow)
- A slider switch (functionally identical to the button-style switch, just with a different sprite)
- Greater variety of sirens, buzzers, bells, whistles, klaxons
- A keypad item; essentially a mini terminal, but tailored to entering short passcodes for doors, cabinets, etc.
- Pull-chain / pull-cord

-  ~~Pressure plates (vertical and horizontal) that activate when pressed / bumped / stepped on~~ <-- motion detectors are already functional equivalents

4.3) Connection Panel Interface
- The connection panel labels for input/output pins that ordinarily transmit binary (or ternary) data should be colour-coded differently from input/output pins that ordinarily transmit alphanumeric data; the labels for binary data could remain blue, whereas those for other data types could be, e.g., white or green

4.4) Tooltips
- Adder component: the min/max clamp tooltips are not worded clearly.  The 'Clamp max' tooltip should read: "The output of the item is restricted to values below this value," and the 'Clamp min' tooltip should read: "The output of the item is restricted to values above this value."

-----
**5.) LIGHTING:**
5.1) Interior Lights and Lamps
- extensible LED strip lighting (linear and curved)
- uplighting / downlighting lamps (essentially wall sconces) that produce columnar, 'v' (i.e., wedge-shaped) and semi-circular beams of light (akin to spotlights / turret lights, but not as powerful); this would allow for a wide variety of dramatic interior lighting, whilst also enabling certain spaces to be lit with lights that would otherwise need to have the 'Cast Shadows' property enabled.
- lights that are part of item sprites should rotate with their parent item

5.2) Visibility and Illumination
- 'Line of Sight' effects are one of the coolest features of Barotrauma, but, as implemented, they are visually distracting and don't always make sense.  Wall segments that are joined together in a light-proof way should not cast 'fog of war' shadows at their seams.  Maybe this could be achieved by allowing wall (and shell) pieces to be linked to each other, like what is already done with hull volumes.  OR shadows could be based on sprite depth.
- Wall caps should cast a complete 'fog of war' shadow across their line of sight.  Right now, duct blocks in particular are not shaded out correctly by wall caps.
- - What *are* wall caps, anyway?  If they are meant to be structural beams that we are viewing in cross-section, then their line-of-sight shadow-casting should behave accordingly.  If they are not, then they shouldn't need to cast shadows separate from whatever is cast by inside walls
- An option to use 'Line of Sight' effects only when outside in the environment, but not while inside outposts or submarines (and vice versa)
- An option for 'Line of Sight' to be room (or hull volume) based, rather than element-based: e.g., everything outside of a given room that is not seen through an opening, a window, or a glass wall is rendered black, as opposed to currently, where individual beams, et cetera cast 'fog of war' shadows.
- Right now, the AI doesn't really use foreground / midground objects or individual structural elements as cover, e.g. for stealth or during gun fights, so the 'Line of Sight' effect is underutilised.

-----
**6.) SHIP WEAPONS:**
6.1) Ship Weapons:
- Water-soluble poisons that could be released (sprayed?), e.g., through ducts or into the open ocean

6.2) Other Exterior-Mounted Installations
- An electrically powered crane or sheer legs that can be used to hoist cargo or to move heavy wreckage on the seafloor.  This could open up new game mechanics and mission types for mid-game, offering more avenues for gated progression: e.g., towing small distressed vessels, salvaging larger wrecks; moving huge sea-floor obstacles

-----
**7.) CREATURES and NPCs:**
7.1) Creatures:
- A wider variety of non-interactable ('flavour') background flora and fauna
- Small, live collectible rare creatures (flora and fauna) and dedicated cages and tanks in which to house / display them; perhaps requiring different baits / lures and methods of capture; these could even be quest items for a new quest type

7.2) NPCs and Bots:
- Crewmember bots' AI should be able to distinguish between buttons / switches that control the opening / closing of doors versus others placed nearby that control other actions; e.g., an override switch that locks a door open or closed should not be something an AI accidentally presses when trying to open a door. This could maybe be solved with:
- - An option in the Submarine Editor item properties interface to make certain items (e.g., buttons, switches, levers) invisible (i.e., non-interactable) to NPCs and AI-controlled crewmembers
- Human NPCs and player characters should be able to catch fire if they stand in a fire long enough; there should be an animation for combustion / burning and a fire damage duration effect
- The current system for health and injuries needs some refinement.  Right now, even minor injuries utilise the same icons and text descriptions as major injuries: a sprained ankle and a broken bone are both advertised the same way, regardless of how far a character tripped or fell.  At the very least, perhaps the injury icons in the bottom right corner of the screen could be colour-coded based on severity of injury

-----
**8.) INVENTORY**
8.1) Player Character Inventory Bar
- As a QoL feature, the first three slots of the inventory bar should be lockable or at the very least assignable to items of a specific kind.  For example, if the 'Revolver' is assigned to slot no. 2, then when a revolver is picked up by the character, it should preferentially fill that slot of the inventory bar if it is free.  Optionally, the revolver could even evict an item that is taking up its assigned slot and shunt it to another available slot.
- - Personally, I would like it if slots 1, 2, 3, and 0 had a special ability to be assignable like this.  Slots locked to a preference could be shaded a different colour than normal slots.

-----
**9.) SUBMARINE EDITOR:**
9.1) General
- The 'Manage Groups' menu should have an option to sort groups alphanumerically by name
- The 'Toggle Visibility' dropdown menu should have an option to toggle visibility by sprite depth
- At the moment, it is not obvious that one may 'SHIFT + RMB' to bring up a menu that includes such useful options as 'Toggle Grid;' this could be mentioned at the end of the short notice to users in the upper right-hand 'Info' (i) pop-up; it would be even better if the 'Toggle Grid' option were also available in the main 'Toggle Visibility' dropdown menu

-----
**10.) GENERAL FORMATTING and UI ISSUES**
10.1) General
- Option to disable 'Out of Ammo' text in-game, or a subtler way of letting players know that their weapons are out of ammo
- Standardised capitalisation of room names when NPCs mention a hull space by its room name; currently, some room names are capitalised even in the middle of NPCs' sentences, but this is inconsistent
- Text localisation should respect capitalisation (e.g., a custom text 'WARNING' (all caps) should not be localised as 'Warning')

-----
**11.) BAROTRAUMA WIKI:**
11.1) General
- There should be a prominently placed link to the 'Submarines' page of the wiki, which is currently not easy to find, and yet is arguably one of the most important hub pages for info

**Comment by Ek-Videogames** [comment]
- More font selections for Text-Display/Label, the default one is terrible low-res.

**Reply by Arctodites** [reply]
Agreed.  A wider selection of fonts (at higher res) with better kerning and distinctive letter-forms to increase legibility (e.g., 8/B, O/0, 1/L/I, m/rn, LI/U)

**Comment by Likvidus** [comment]
There are a lot of good ideas here, +1
